activerecord/lib/active_record/reflection.rb

```diff
@@ -466,7 +466,12 @@ def join_table
       end
 
       def foreign_key
-        @foreign_key ||= -(options[:foreign_key]&.to_s || derive_foreign_key)
+        @foreign_key ||= if options[:foreign_key] && options[:foreign_key].is_a?(Array)
```

Comment:
> to take a symbol or array of symbols and then internally automatically and immediately turn those into arrays

This is basically the final goal. At the end `foreign_key` and `active_record_primary_key` methods should be the only ones that use `Array.wrap` to support both symbol and array of symbols as a value and everything else downstream should expect to work with an array. 
However doing this right now means growing this PR and this is what I was trying to avoid. For example we will have to additionally change places that still expect `foreign_key` to be a single string
For example:

https://github.com/rails/rails/blob/2e28ae403d9247181220b8a79273f953a892df69/activerecord/lib/active_record/autosave_association.rb#L451
https://github.com/rails/rails/blob/2e28ae403d9247181220b8a79273f953a892df69/activerecord/lib/active_record/autosave_association.rb#L474
https://github.com/rails/rails/blob/2e28ae403d9247181220b8a79273f953a892df69/activerecord/lib/active_record/relation.rb#L47
https://github.com/rails/rails/blob/2e28ae403d9247181220b8a79273f953a892df69/activerecord/lib/active_record/associations/belongs_to_association.rb#L125

https://github.com/Shopify/rails/blob/2e28ae403d9247181220b8a79273f953a892df69/activerecord/lib/active_record/associations/belongs_to_association.rb#L134

I wonder if there is an option for us to somehow avoid calling `is_a?` more than once per application boot as I don't think we expect `foreign_key` value to change once configured and model is loaded. Might be an overkill though

Reply:
>I wonder if there is an option for us to somehow avoid calling is_a? more than once per application boot 

I took another look and somehow I missed the fact that we actually never call `foreign_key` more than once. So only the first query will hit `.is_a?` check and then the value is going to be memoized which is unlikely a noticeable regression. Let me know if I might have missed some other use-cases but a simple usage as 
```ruby
    post.comments.to_a
    post.comments.to_a
    Post.first.comments
    Post.first.comments
    Post.includes(:comments).all
    Post.includes(:comments).all
```
Allocates `HasManyReflection @name=:comments` once and memoizes `foreign_key` value on it

Reply:
Sure, here are the results of running the code like:
```ruby
post = Post.create!

comment = Comment.create!(post: post)
Comment.create!(post: post)
Comment.create!(post: post)

Benchmark.ips do |x|
  x.report("main") do
    post.comments.to_a
    post.association(:comments).reset
    comment.post
    comment.association(:post).reset
  end
end
```
Against `main` and the proposed PR:

```
main:
883.013  (±19.1%) i/s -      4.214k in   5.003156s
927.900  (± 4.8%) i/s -      4.704k in   5.081965s
938.130  (± 4.2%) i/s -      4.704k in   5.023328s

Current PR:
985.325  (± 4.1%) i/s -      4.984k in   5.067057s
970.700  (± 3.5%) i/s -      4.900k in   5.054164s
951.268  (± 3.5%) i/s -      4.770k in   5.020796s
```

The difference is basically not visible as other parts of the querying flow make much more time than a single `is_a?` check takes. Especially considering that we are not doing most of the `is_a?` checks more than once. It's all memoized once the reflection is defined. 

Though it is still a reasonable concern to keep in mind as this is not all the checks we are going to add. So I'm going to keep the benchmark to make sure none of the backwards compatibility checks cause a regression

Reply:
I went ahead and ran benchmarks against another branch I have with more changes and more `Array.wrap` and `is_a?` checks and it's still not noticeable:

```
984.023  (± 4.3%) i/s -      4.950k in   5.039984s
987.507  (± 5.2%) i/s -      4.998k in   5.076183s
973.145  (± 4.4%) i/s -      4.949k in   5.096015s
```